File: benchmarks/consumer.py
#!/usr/bin/env python3
"""
Consumer for benchmarking: receives random bytes via STDIO or shared memory.
"""

import argparse
import os
import struct
import sys
import time
from typing import Optional

# Add parent directory to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'python'))

from ipc0cp.ring_buffer import SharedRingBufferConsumer
from ipc0cp import IPCException


def consume_stdio() -> dict:
    """
    Consume random bytes via STDIN (binary mode) until EOF.
    
    Protocol: [length:8 bytes][payload:length bytes]
        
    Returns:
        Statistics dict with bytes_received, messages_received, elapsed_time
    """
    # Use binary stdin
    stdin = sys.stdin.buffer
    
    bytes_received = 0
    messages_received = 0
    start_time = time.time()
    
    while True:
        # Read length prefix (8 bytes)
        length_bytes = stdin.read(8)
        if len(length_bytes) < 8:
            # EOF or incomplete read
            break
        
        size = struct.unpack('<Q', length_bytes)[0]
        
        # Check for end-of-stream marker
        if size == 0:
            # Producer signaled end of stream
            break
        
        # Read payload
        payload = stdin.read(size)
        if len(payload) < size:
            # Incomplete read
            break
        
        bytes_received += size
        messages_received += 1
    
    elapsed_time = time.time() - start_time
    
    return {
        'bytes_received': bytes_received,
        'messages_received': messages_received,
        'elapsed_time': elapsed_time,
        'throughput_mbps': (bytes_received / elapsed_time) / (1024 * 1024),
    }


def consume_shm(shm_name: str) -> dict:
    """
    Consume random bytes via shared memory ring buffer until empty.
    
    Args:
        shm_name: Shared memory segment name
        
    Returns:
        Statistics dict with bytes_received, messages_received, elapsed_time
    """
    # Wait a bit for producer to create shared memory
    time.sleep(0.5)
    
    # Create consumer with 2GB buffer
    consumer = SharedRingBufferConsumer(
        shm_name=shm_name,
        total_data_bytes=2 * 1024 * 1024 * 1024,  # 2 GB
        blocking=True,
        auto_attach=True,
    )
    
    bytes_received = 0
    messages_received = 0
    start_time = time.time()
    
    try:
        while True:
            # Pop from ring buffer (blocking with no timeout)
            # Returns None only for end-of-stream, throws exception for errors
            try:
                payload = consumer.pop(timeout=None)
            except IPCException as e:
                # Handle exceptions - these are real errors, not EOS
                print(f"Error during pop: {e} (error_type={e.error_type})", file=sys.stderr)
                raise
            
            if payload is None:
                # End-of-stream received - clean exit
                break
            
            # Normal data
            size = len(payload)
            bytes_received += size
            messages_received += 1
        
        elapsed_time = time.time() - start_time
        
        return {
            'bytes_received': bytes_received,
            'messages_received': messages_received,
            'elapsed_time': elapsed_time,
            'throughput_mbps': (bytes_received / elapsed_time) / (1024 * 1024),
        }
    finally:
        consumer.close()


def main():
    parser = argparse.ArgumentParser(description='Benchmark consumer')
    
    # Transport mode
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument('--stdio', action='store_true', help='Use STDIN/STDOUT')
    group.add_argument('--shm', type=str, help='Shared memory segment name')
    
    # Benchmark parameters
    parser.add_argument('--quiet', action='store_true',
                        help='Suppress progress output')
    
    args = parser.parse_args()
    
    if not args.quiet:
        print(f"Consumer starting...", file=sys.stderr)
    
    if args.stdio:
        stats = consume_stdio()
    else:
        stats = consume_shm(args.shm)
    
    # Print stats to stderr
    print(f"\nConsumer Stats:", file=sys.stderr)
    print(f"  Bytes received: {stats['bytes_received']:,} ({stats['bytes_received'] / (1024**3):.2f} GB)", file=sys.stderr)
    print(f"  Messages received: {stats['messages_received']:,}", file=sys.stderr)
    print(f"  Elapsed time: {stats['elapsed_time']:.2f} seconds", file=sys.stderr)
    print(f"  Throughput: {stats['throughput_mbps']:.2f} MB/s", file=sys.stderr)


if __name__ == '__main__':
    main()

File: benchmarks/consumer.cpp
#include "ipc0cp/ring_buffer.hpp"
#include "ipc0cp/stdio.hpp"

#include <chrono>
#include <iostream>
#include <string>
#include <string_view>
#include <thread>

using namespace ipc0cp;

namespace {

struct Config {
    bool use_stdio = false;
    bool use_shm = false;
    std::string shm_name;

    bool quiet = false;
    size_t shm_total_data_bytes = 2ULL * 1024 * 1024 * 1024;

    std::chrono::seconds attach_timeout{5};
};

bool parse_args(int argc, char** argv, Config& config) {
    for (int i = 1; i < argc; ++i) {
        std::string_view arg(argv[i]);
        if (arg == "--stdio") {
            config.use_stdio = true;
        } else if (arg == "--shm" && i + 1 < argc) {
            config.use_shm = true;
            config.shm_name = std::string(argv[++i]);
        } else if (arg == "--quiet") {
            config.quiet = true;
        }
    }

    if (config.use_stdio == config.use_shm) {
        std::cerr << "Exactly one of --stdio or --shm <name> is required" << std::endl;
        return false;
    }

    if (config.use_shm && config.shm_name.empty()) {
        std::cerr << "Missing shm name for --shm" << std::endl;
        return false;
    }

    return true;
}

size_t payload_size_from_obj(const RingBufferObject& obj) {
    if (!obj.data) {
        return 0;
    }
    if (auto* bytes_obj = dynamic_cast<BytesData*>(obj.data.get())) {
        return bytes_obj->bytes.size();
    }
    auto serialized = obj.data->serialize();
    return serialized.payload.size();
}

}  // namespace

int main(int argc, char** argv) {
    Config config;
    if (!parse_args(argc, argv, config)) {
        return 2;
    }

    size_t bytes_received = 0;
    size_t messages = 0;
    auto start = std::chrono::steady_clock::now();

    if (config.use_stdio) {
        StdioConsumer consumer;
        while (true) {
            auto result = consumer.pop_raw();
            if (!result) {
                break;
            }
            bytes_received += result->second.size();
            ++messages;
        }
    } else {
        // Allow producer a short time to create the shared memory
        std::this_thread::sleep_for(std::chrono::milliseconds(100));

        SharedRingBufferConsumer consumer(
            config.shm_name,
            config.shm_total_data_bytes,
            true,
            false,
            true
        );

        const auto attach_deadline = std::chrono::steady_clock::now() + config.attach_timeout;
        while (true) {
            if (consumer.attach()) {
                break;
            }

            const auto err = consumer.get_last_error();
            if (err == IPCError::ShmNotFound && std::chrono::steady_clock::now() < attach_deadline) {
                std::this_thread::sleep_for(std::chrono::milliseconds(10));
                continue;
            }

            std::cerr << "Failed to attach to shared memory: " << errorToString(err) << std::endl;
            return 1;
        }

        try {
            while (true) {
                auto obj = consumer.pop();
                if (!obj) {
                    break;
                }
                bytes_received += payload_size_from_obj(*obj);
                ++messages;
            }
        } catch (...) {
            consumer.close();
            throw;
        }

        consumer.close();
    }

    const auto elapsed = std::chrono::duration<double>(
        std::chrono::steady_clock::now() - start
    ).count();

    const double throughput = elapsed > 0
        ? (static_cast<double>(bytes_received) / (1024.0 * 1024.0)) / elapsed
        : 0.0;

    std::cerr << "\nConsumer Stats:\n";
    std::cerr << "  Bytes received (payload only): " << bytes_received << "\n";
    std::cerr << "  Messages received: " << messages << "\n";
    std::cerr << "  Elapsed time: " << elapsed << " seconds\n";
    std::cerr << "  Throughput: " << throughput << " MB/s\n";

    return 0;
}
